beacon_node/beacon_chain/src/data_availability_checker/processing_cache.rs

```diff
@@ -0,0 +1,75 @@
+use crate::data_availability_checker::AvailabilityView;
+use std::collections::hash_map::Entry;
+use std::collections::HashMap;
+use types::beacon_block_body::{KzgCommitmentOpts, KzgCommitments};
+use types::{EthSpec, Hash256, Slot};
+
+/// This cache is used only for gossip blocks/blobs and single block/blob lookups, to give req/resp
+/// a view of what we have and what we require. This cache serves a slightly different purpose than
+/// gossip caches because it allows us to process duplicate blobs that are valid in gossip.
+/// See `AvailabilityView`'s trait definition.
+#[derive(Default)]
+pub struct ProcessingCache<E: EthSpec> {
+    processing_cache: HashMap<Hash256, ProcessingComponents<E>>,
+}
+
+impl<E: EthSpec> ProcessingCache<E> {
+    pub fn get(&self, block_root: &Hash256) -> Option<&ProcessingComponents<E>> {
+        self.processing_cache.get(block_root)
+    }
+    pub fn entry(&mut self, block_root: Hash256) -> Entry<'_, Hash256, ProcessingComponents<E>> {
+        self.processing_cache.entry(block_root)
+    }
+    pub fn remove(&mut self, block_root: &Hash256) {
+        self.processing_cache.remove(block_root);
+    }
+    pub fn has_block(&self, block_root: &Hash256) -> bool {
+        self.processing_cache
+            .get(block_root)
+            .map_or(false, |b| b.block_exists())
+    }
+    pub fn incomplete_processing_components(&self, slot: Slot) -> Vec<Hash256> {
+        let mut roots_missing_components = vec![];
+        for (&block_root, info) in self.processing_cache.iter() {
+            if info.slot == slot && !info.is_available() {
+                roots_missing_components.push(block_root);
+            }
+        }
+        roots_missing_components
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct ProcessingComponents<E: EthSpec> {
+    slot: Slot,
+    /// Blobs required for a block can only be known if we have seen the block. So `Some` here
+    /// means we've seen it, a `None` means we haven't. The `kzg_commitments` value helps us figure
+    /// out whether incoming blobs actually match the block.
+    pub kzg_commitments: Option<KzgCommitments<E>>,
+    /// On insertion, a collision at an index here when `required_blobs` is
```

Comment:
Do you mean `kzg_commitments` instead of `required_blobs` here?
Also have a naming suggestion here:
`kzg_commitments -> block_commitments`
`processing_blobs -> blob_commitments`
A little more self explanatory what the fields mean imo

Reply:
Oh shoot this comment is out of date, I was initially thinking we'd want to create a new entry on a collision for this particular cache, but upon further thinking I don't think we want/need that and instead should prefer the `AvailabilityView` logic. Thinking about this sort of logic over three different caches is part of what motivated the trait 😅 

great rename suggestions!

Reply:
updates: https://github.com/sigp/lighthouse/pull/4732/commits/49b0c2bf146d8f3e6de1d83c861e060a34ba9e25